// react-native-label/dist/components/Banner.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactNative = require("react-native");

var _typeCheck = require("type-check");

var _ = require(".");

var _constants = require("../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var hyp = function hyp(a, b) {
  return Math.sqrt(Math.pow(a, 2) + Math.pow(b, 2));
};

var defaultStyle = Object.freeze({
  fontSize: 30,
  color: 'white',
  textAlign: 'center',
  alignItems: 'center',
  justifyContent: 'center'
});

var createStyles = function createStyles(layout, distance) {
  var _StyleSheet$create;

  var width = layout.width,
      height = layout.height;
  var baseProps = {
    width: width,
    height: height
  };
  return _reactNative.StyleSheet.create((_StyleSheet$create = {}, _defineProperty(_StyleSheet$create, _constants.Orientation.TOP_LEFT, _objectSpread(_objectSpread({}, baseProps), {}, {
    transform: [{
      translateX: -0.5 * width
    }, {
      translateY: -0.5 * height
    }, {
      rotate: '-45deg'
    }, {
      translateY: 0.5 * (height + distance)
    }]
  })), _defineProperty(_StyleSheet$create, _constants.Orientation.TOP_RIGHT, _objectSpread(_objectSpread({}, baseProps), {}, {
    transform: [{
      translateX: 0.5 * width
    }, {
      translateY: -0.5 * height
    }, {
      rotate: '45deg'
    }, {
      translateY: 0.5 * (height + distance)
    }]
  })), _defineProperty(_StyleSheet$create, _constants.Orientation.BOTTOM_LEFT, _objectSpread(_objectSpread({}, baseProps), {}, {
    justifyContent: 'flex-end',
    transform: [{
      translateX: -0.5 * width
    }, {
      translateY: 0.5 * height
    }, {
      rotate: '45deg'
    }, {
      translateY: -0.5 * (height + distance)
    }]
  })), _defineProperty(_StyleSheet$create, _constants.Orientation.BOTTOM_RIGHT, _objectSpread(_objectSpread({}, baseProps), {}, {
    justifyContent: 'flex-end',
    transform: [{
      translateX: 0.5 * width
    }, {
      translateY: 0.5 * height
    }, {
      rotate: '-45deg'
    }, {
      translateY: -0.5 * (height + distance)
    }]
  })), _StyleSheet$create));
};

var createEnterStyles = function createEnterStyles(layout, _ref, distance) {
  var _StyleSheet$create2;

  var height = _ref.height;
  var baseProps = {
    position: "absolute"
  };
  var dist = hyp(distance, distance) * 0.5;
  var dim = hyp(height, height) * -1;
  return _reactNative.StyleSheet.create((_StyleSheet$create2 = {}, _defineProperty(_StyleSheet$create2, _constants.Orientation.TOP_LEFT, _objectSpread(_objectSpread({}, baseProps), {}, {
    transform: [{
      rotate: '270deg'
    }],
    top: dist,
    left: dim
  })), _defineProperty(_StyleSheet$create2, _constants.Orientation.TOP_RIGHT, _objectSpread(_objectSpread({}, baseProps), {}, {
    top: dim,
    right: dist
  })), _defineProperty(_StyleSheet$create2, _constants.Orientation.BOTTOM_LEFT, _objectSpread(_objectSpread({}, baseProps), {}, {
    bottom: dim,
    left: dist,
    transform: [{
      rotate: '180deg'
    }]
  })), _defineProperty(_StyleSheet$create2, _constants.Orientation.BOTTOM_RIGHT, _objectSpread(_objectSpread({}, baseProps), {}, {
    bottom: dim,
    right: dist,
    transform: [{
      scale: -1
    }]
  })), _StyleSheet$create2));
};

var createExitStyles = function createExitStyles(layout, _ref2, distance) {
  var _StyleSheet$create3;

  var height = _ref2.height;
  var baseProps = {
    position: "absolute"
  };
  var dist = hyp(distance, distance) * 0.5;
  var dim = hyp(height, height) * -1;
  return _reactNative.StyleSheet.create((_StyleSheet$create3 = {}, _defineProperty(_StyleSheet$create3, _constants.Orientation.TOP_LEFT, _objectSpread(_objectSpread({}, baseProps), {}, {
    left: dist,
    top: dim
  })), _defineProperty(_StyleSheet$create3, _constants.Orientation.TOP_RIGHT, _objectSpread(_objectSpread({}, baseProps), {}, {
    transform: [{
      rotate: '90deg'
    }],
    right: dim,
    top: dist
  })), _defineProperty(_StyleSheet$create3, _constants.Orientation.BOTTOM_LEFT, _objectSpread(_objectSpread({}, baseProps), {}, {
    transform: [{
      rotate: '270deg'
    }],
    left: dim,
    bottom: dist
  })), _defineProperty(_StyleSheet$create3, _constants.Orientation.BOTTOM_RIGHT, _objectSpread(_objectSpread({}, baseProps), {}, {
    transform: [{
      rotate: '90deg'
    }],
    right: dim,
    bottom: dist
  })), _StyleSheet$create3));
};

var Banner = function Banner(_ref3) {
  var orientation = _ref3.orientation,
      distance = _ref3.distance,
      layout = _ref3.layout,
      shadowProps = _ref3.shadowProps,
      title = _ref3.title,
      color = _ref3.color,
      extraStyle = _ref3.style,
      ratio = _ref3.ratio,
      extent = _ref3.extent,
      extraProps = _objectWithoutProperties(_ref3, ["orientation", "distance", "layout", "shadowProps", "title", "color", "style", "ratio", "extent"]);

  var _useState = (0, _react.useState)(null),
      _useState2 = _slicedToArray(_useState, 2),
      childLayout = _useState2[0],
      setChildLayout = _useState2[1];

  var _createStyles = createStyles(layout, distance),
      style = _createStyles[orientation];

  var w = layout.width,
      h = layout.height;
  var width = Math.sqrt(Math.pow(layout.width, 2) + Math.pow(layout.height, 2));
  var Component = (0, _typeCheck.typeCheck)("Function", title) && title;
  var childStyle = [_objectSpread(_objectSpread({}, defaultStyle), extraStyle), shadowProps, {
    width: width,
    backgroundColor: color
  }];
  var onLayout = (0, _react.useCallback)(function (_ref4) {
    var layout = _ref4.nativeEvent.layout;
    return setChildLayout(layout);
  }, [setChildLayout]);
  return /*#__PURE__*/_react["default"].createElement(_reactNative.Animated.View, {
    pointerEvents: "none",
    key: orientation,
    style: {
      flex: 1
    }
  }, (0, _typeCheck.typeCheck)("Object", childLayout) && /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(_.Tail, {
    style: createEnterStyles(layout, childLayout, distance)[orientation],
    width: hyp(childLayout.height, childLayout.height),
    height: hyp(childLayout.height, childLayout.height),
    color: color,
    ratio: ratio,
    extent: extent
  }), /*#__PURE__*/_react["default"].createElement(_.Tail, {
    style: createExitStyles(layout, childLayout, distance)[orientation],
    width: hyp(childLayout.height, childLayout.height),
    height: hyp(childLayout.height, childLayout.height),
    color: color,
    ratio: ratio,
    extent: extent
  })), /*#__PURE__*/_react["default"].createElement(_reactNative.Animated.View, {
    pointerEvents: "box-none",
    style: {
      flex: 1,
      overflow: 'hidden'
    }
  }, /*#__PURE__*/_react["default"].createElement(_reactNative.Animated.View, {
    pointerEvents: "box-none",
    style: [style, {
      alignItems: 'center',
      opacity: !!childLayout ? 1 : 0
    }]
  }, (0, _typeCheck.typeCheck)("String", title) && /*#__PURE__*/_react["default"].createElement(_reactNative.Animated.Text, {
    onLayout: onLayout,
    pointerEvents: "none",
    style: childStyle,
    children: title
  }), !!Component && /*#__PURE__*/_react["default"].createElement(_reactNative.Animated.View, {
    onLayout: onLayout,
    pointerEvents: "box-none",
    style: childStyle
  }, /*#__PURE__*/_react["default"].createElement(Component, extraProps)))));
};

Banner.propTypes = {
  orientation: _propTypes["default"].oneOf(Object.values(_constants.Orientation)).isRequired,
  distance: _propTypes["default"].number.isRequired,
  shadowProps: _propTypes["default"].shape({}),
  style: _propTypes["default"].shape({}),
  color: _propTypes["default"].string,
  ratio: _propTypes["default"].number,
  extent: _propTypes["default"].number
};
Banner.defaultProps = {
  shadowProps: {
    shadowColor: "#000",
    shadowOffset: {
      width: 0,
      height: 12
    },
    shadowOpacity: 0.2,
    shadowRadius: 8,
    elevation: 24
  },
  style: defaultStyle,
  color: "#C2185B",
  ratio: 0.8,
  extent: 0.5
};

var _default = /*#__PURE__*/_react["default"].memo(Banner);

exports["default"] = _default;
