// react-native-label/dist/components/Tail.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactNative = require("react-native");

var _ = require(".");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var styles = _reactNative.StyleSheet.create({
  peaks: {
    width: "100%",
    flexDirection: "row"
  },
  gap: {
    position: "absolute",
    bottom: 0,
    height: 1
  }
});

var Tail = function Tail(_ref) {
  var style = _ref.style,
      width = _ref.width,
      height = _ref.height,
      color = _ref.color,
      ratio = _ref.ratio,
      extent = _ref.extent,
      shadowProps = _ref.shadowProps,
      extraProps = _objectWithoutProperties(_ref, ["style", "width", "height", "color", "ratio", "extent", "shadowProps"]);

  return /*#__PURE__*/_react["default"].createElement(_reactNative.View, {
    style: [style, {
      width: width,
      height: height
    }]
  }, /*#__PURE__*/_react["default"].createElement(_reactNative.View, {
    style: {
      height: (1 - extent) * height
    }
  }), /*#__PURE__*/_react["default"].createElement(_reactNative.View, {
    style: {
      width: width,
      height: extent * height
    }
  }, /*#__PURE__*/_react["default"].createElement(_reactNative.View, {
    style: styles.peaks
  }, /*#__PURE__*/_react["default"].createElement(_.Triangle, {
    width: width * 0.5,
    height: Math.ceil(height * ratio * extent),
    color: color
  }), /*#__PURE__*/_react["default"].createElement(_.Triangle, {
    width: width * 0.5,
    height: Math.ceil(height * ratio * extent),
    color: color
  })), /*#__PURE__*/_react["default"].createElement(_reactNative.View, {
    style: shadowProps
  }, /*#__PURE__*/_react["default"].createElement(_reactNative.View, {
    style: {
      width: width,
      height: height * (1 - ratio) * extent,
      backgroundColor: color
    }
  }))));
};

Tail.propTypes = {
  width: _propTypes["default"].number,
  height: _propTypes["default"].number,
  color: _propTypes["default"].string,
  ratio: _propTypes["default"].number,
  extent: _propTypes["default"].number,
  shadowProps: _propTypes["default"].shape({})
};
Tail.defaultProps = {
  width: 100,
  height: 100,
  color: "orange",
  ratio: 0.5,
  extent: 1,
  shadowProps: {
    shadowColor: "black",
    shadowOffset: {
      width: 0,
      height: 0
    },
    shadowOpacity: 0.2,
    shadowRadius: 5,
    elevation: 48
  }
};

var _default = /*#__PURE__*/_react["default"].memo(Tail);

exports["default"] = _default;
